backend/src/v2/driver/driver.go

```diff
@@ -671,13 +671,26 @@ func resolveInputs(ctx context.Context, dag *metadata.DAG, iterationIndex *int,
 				}
 			}
 			value, hasValue := inputs.GetParameterValues()[name]
-			if !hasValue {
+
+			// Handle when parameter does not have input value
+			if !hasValue && inputsSpec.GetParameters()[name].IsOptional == false {
+				// when parameter is not optional, value comes from default value, otherwise error
 				if spec.GetDefaultValue() == nil {
-					return fmt.Errorf("input parameter %q is required", name)
+					return fmt.Errorf("no value or default value provided for non-optional parameter %q", name)
 				}
 				inputs.GetParameterValues()[name] = spec.GetDefaultValue()
```

Comment:
I think this is probably dead code. If a parameter has a default value then optional must be true. 
"no value + not optional" goes to error directly.

Reply:
This is true. I will remove this.